bigquery_etl/format_sql/formatter.py

```diff
@@ -0,0 +1,189 @@
+            # decrease indent to top of current block
+            while indent_types and indent_types[-1] is TopLevelKeyword:
+                indent_types.pop()
+
+        # yield whitespace
+        if first_token or isinstance(token, Comment):
+            pass  # no space before first token and comments contain original whitespace
+        elif require_newline_before_next_token or isinstance(
+            token, (NewlineKeyword, ClosingBracket)
+        ):
+            yield Whitespace("\n" + indent * len(indent_types))
+        elif (
+            allow_space_before_next_token
+            and (
+                allow_space_before_next_bracket or not isinstance(token, OpeningBracket)
+            )
+            and not isinstance(
+                token, (NoSpaceOperator, ExpressionSeparator, StatementSeparator)
+            )
+        ):
+            yield Whitespace(" ")
+
+        yield token
+
+        # update state for next token
+        require_newline_before_next_token = isinstance(
+            token,
+            (
+                Comment,
+                TopLevelKeyword,
+                OpeningBracket,
+                ExpressionSeparator,
+                StatementSeparator,
+            ),
+        )
+        allow_space_before_next_token = not isinstance(token, NoSpaceOperator)
+        allow_space_before_next_bracket = isinstance(
+            token, (SpaceBeforeBracketKeyword, Operator)
+        )
+        if isinstance(token, (TopLevelKeyword, OpeningBracket)):
+            # increase indent
+            indent_types.append(type(token))
+        elif isinstance(token, StatementSeparator):
+            # reset indent
+            indent_types = []
+        first_token = False
+
+
+class Line:
+    """Container for a line of tokens."""
+
+    def __init__(self, indent_token=None):
+        """Initialize."""
+        self.indent_token = indent_token
+        if indent_token is None:
+            self.indent_level = 0
+        else:
+            self.indent_level = len(indent_token.value)
+            if indent_token.value.startswith("\n"):
+                self.indent_level -= 1
+        self.inline_tokens = []
+        self.inline_length = 0
+        self.contains_comment = isinstance(indent_token, Comment)
+        self.can_start_inline_block = False
+
+    def add(self, token):
+        """Add a token to this line."""
+        self.inline_length += len(token.value)
+        self.inline_tokens.append(token)
+        self.contains_comment = self.contains_comment or isinstance(token, Comment)
+        self.can_start_inline_block = isinstance(token, OpeningBracket)
+
+    @property
+    def tokens(self):
+        """Get a list of all the tokens in this line."""
+        if self.indent_token is None:
+            return self.inline_tokens
+        else:
+            return [self.indent_token] + self.inline_tokens
+
+
+def inline_block_format(tokens, max_inline_length=100):
+    """Extend simple_format with inline blocks.
```

Comment:
What is an inline block?

Reply:
I'm open to suggestions on wording, I came up with this because the first docstring line has to fit within 88 characters.

> What is an inline block?

a bracket block, i.e. anything wrapped in brackets, that is on a single line instead of multiple lines.


for example, reducing:
```sql
IF(
  condition,
  value_if_true,
  value_if_false
)
```
to
```sql
IF(condition, value_if_true, value_if_false)
```
but only when it doesn't exceed `max_line_length`.

Reply:
The wording is fine, this example would clarifies the wording quite a bit. It would be useful to add below the next sentence.